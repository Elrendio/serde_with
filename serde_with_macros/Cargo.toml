[package]
name = "serde_with_macros"
version = "1.0.1"
authors = ["jonasbb"]

description = "proc-macro library for serde_with"
documentation = "https://docs.rs/serde_with_macros/"
repository = "https://github.com/jonasbb/serde_with/serde_with_macros"
readme = "README.md"
keywords = ["serde", "utilities", "serialization", "deserialization"]
license = "MIT OR Apache-2.0"

[lib]
proc-macro = true

[badges]
travis-ci = { repository = "jonasbb/serde_with", branch = "master" }
codecov = { repository = "jonasbb/serde_with", branch = "master", service = "github" }
maintenance = { status = "actively-developed" }

[dependencies]
proc-macro2 = "1.0.1"
quote = "1.0.0"

[dependencies.syn]
version = "1.0.0"
features = [
    # "extra-traits", # Only for debugging
    "full",
]

[dev-dependencies]
pretty_assertions = "0.5.1"
rustc_version = "0.2.3"
serde = { version = "1.0.75", features = [ "derive" ] }
serde_json = "1.0.25"
trybuild = "1.0.4"
version-sync = "0.7.0"

[package.metadata.docs.rs]
all-features = true
