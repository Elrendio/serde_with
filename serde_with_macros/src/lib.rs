#![deny(
    missing_debug_implementations,
    missing_copy_implementations,
    trivial_casts,
    trivial_numeric_casts,
    unused_extern_crates,
    unused_import_braces,
    unused_qualifications,
    variant_size_differences
)]
#![cfg_attr(feature = "cargo-clippy", allow(renamed_and_removed_lints))]
#![doc(html_root_url = "https://docs.rs/serde_with_macros/1.0.0")]

extern crate proc_macro;
extern crate proc_macro2;
extern crate quote;
extern crate syn;

use proc_macro::TokenStream;
use proc_macro2::Span;
use quote::quote;
use syn::{parse::Parser, Attribute, Error, Field, Fields, ItemEnum, ItemStruct, Path, Type};

/// Return `true`, if the type path refers to `std::option::Option`
///
/// Accepts
///
/// * `Option`
/// * `std::option::Option`, with or without leading `::`
/// * `core::option::Option`, with or without leading `::`
fn is_std_option(path: &Path) -> bool {
    (path.leading_colon.is_none() && path.segments.len() == 1 && path.segments[0].ident == "Option")
        || (path.segments.len() == 3
            && (path.segments[0].ident == "std" || path.segments[0].ident == "core")
            && path.segments[1].ident == "option"
            && path.segments[2].ident == "Option")
}

/// Add the skip_serializing_if annotation to each field of the struct
fn skip_serializing_null_add_attr_to_field<'a>(fields: impl IntoIterator<Item = &'a mut Field>) {
    fields.into_iter().for_each(|field| {
        if let Type::Path(path) = &field.ty {
            if is_std_option(&path.path) {
                // FIXME if skip_serializing_if already exists, do not add it again
                // Add the `skip_serializing_if` attribute
                let attr_tokens = quote!(
                    #[serde(skip_serializing_if = "Option::is_none")]
                );
                let parser = Attribute::parse_outer;
                let attrs = parser
                    .parse2(attr_tokens)
                    .expect("Static attr tokens should not panic");
                field.attrs.extend(attrs);
            }
        }
    })
}

/// Handle a single struct or a single enum variant
fn skip_serializing_null_handle_fields(fields: &mut Fields) {
    match fields {
        // simple, no fields, do nothing
        Fields::Unit => {}
        Fields::Named(ref mut fields) => {
            skip_serializing_null_add_attr_to_field(fields.named.iter_mut())
        }
        Fields::Unnamed(ref mut fields) => {
            skip_serializing_null_add_attr_to_field(fields.unnamed.iter_mut())
        }
    };
}

#[proc_macro_attribute]
pub fn skip_serializing_null(_args: TokenStream, input: TokenStream) -> TokenStream {
    // For each field in the struct given by `input`, add the `skip_serializing_if` attribute,
    // if and only if, it is of type `Option`
    let res = if let Ok(mut input) = syn::parse::<ItemStruct>(input.clone()) {
        skip_serializing_null_handle_fields(&mut input.fields);
        quote!(#input)
    } else if let Ok(mut input) = syn::parse::<ItemEnum>(input.clone()) {
        input.variants.iter_mut().for_each(|variant| {
            skip_serializing_null_handle_fields(&mut variant.fields);
        });
        quote!(#input)
    } else {
        let span = Span::call_site();
        Error::new(
            span,
            "The attribute can only be applied to struct or enum definitions.",
        )
        .to_compile_error()
    };
    // Hand the modified input back to the compiler
    TokenStream::from(res)
}
