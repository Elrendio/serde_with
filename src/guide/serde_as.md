# `serde_as` Annotation

This is an alternative to serde's with-annotation.
It is more flexible and composable but work with fewer types.

The scheme is based on two new traits, [`SerializeAs`] and [`DeserializeAs`], which need to be implemented by all types which want to be compatible with `serde_as`.
The proc macro attribute [`#[serde_as]`][crate::serde_as] exists as a usability boost for users.

This site contains some general advice how to use this crate and then lists the implemented conversions for `serde_as`.
The basic design of the system was done by [@markazmierczak](https://github.com/markazmierczak).

1. [Switching from serde's with to `serde_as`](#switching-from-serdes-with-to-serde_as)
    1. [Deserializing Optional Fields](#deserializing-optional-fields)
    2. [Implementing `SerializeAs` / `DeserializeAs`](#implementing-serializeas--deserializeas)
    3. [Using `serde_as` in a procedural macro](#re-exporting-serde_as)
2. [De/Serialize Implementations Available](#deserialize-implementations-available)
    1. [Bytes / `Vec<u8>` to hex string](#bytes--vecu8-to-hex-string)
    2. [`Default` from `null`](#default-from-null)
    3. [De/Serialize with `FromStr` and `Display`](#deserialize-with-fromstr-and-display)
    4. [`Duration` as seconds](#duration-as-seconds)
    5. [Ignore deserialization errors](#ignore-deserialization-errors)
    6. [`Maps` to `Vec` of tuples](#maps-to-vec-of-tuples)
    7. [`NaiveDateTime` like UTC timestamp](#naivedatetime-like-utc-timestamp)
    8. [`None` as empty `String`](#none-as-empty-string)
    9. [Timestamps as seconds since UNIX epoch](#timestamps-as-seconds-since-unix-epoch)
    10. [Value into JSON String](#value-into-json-string)
    11. [`Vec` of tuples to `Maps`](#vec-of-tuples-to-maps)

## Switching from serde's with to `serde_as`

For the user the main difference is that instead of

```rust,ignore
#[serde(with = "...")]
```

you now have to write

```rust,ignore
#[serde_as(as = "...")]
```

and place the `#[serde_as]` attribute *before* the `#[derive]` attribute.
You still need the `#[derive(Serialize, Deserialize)]` on the struct/enum.

All together this looks like:

```rust
use serde::{Deserialize, Serialize};
use serde_with::{serde_as, DisplayFromStr};

#[serde_as]
#[derive(Serialize, Deserialize)]
struct A {
    #[serde_as(as = "DisplayFromStr")]
    mime: mime::Mime,
}
```

The main advantage is that you can compose `serde_as` stuff, which is not possible with the with-annotation.
For example, the `mime` field from above could be nested in one or more data structures:

```rust
# use std::collections::BTreeMap;
# use serde::{Deserialize, Serialize};
# use serde_with::{serde_as, DisplayFromStr};
#
#[serde_as]
#[derive(Serialize, Deserialize)]
struct A {
    #[serde_as(as = "Option<BTreeMap<_, Vec<DisplayFromStr>>>")]
    mime: Option<BTreeMap<String, Vec<mime::Mime>>>,
}
```

### Deserializing Optional Fields

During deserialization serde treats fields of `Option<T>` as optional and does not require them to be present.
This breaks when adding either the `serde_as` annotation or serde's `with` annotation.
The default behavior can be restored by adding serde's `default` attribute.

```rust
# use serde::{Deserialize, Serialize};
# use serde_with::{serde_as, DisplayFromStr};
#
#[serde_as]
#[derive(Serialize, Deserialize)]
struct A {
    #[serde_as(as = "Option<DisplayFromStr>")]
    // Allows deserialization without providing a value for `val`
    #[serde(default)]
    val: Option<u32>,
}
```

In the future this behavior might change and `default` would be applied on `Option<T>` fields.
You can add your feedback at [serde_with#185].

### Implementing `SerializeAs` / `DeserializeAs`

You can support [`SerializeAs`] / [`DeserializeAs`] on your own types too.
Most "leaf" types do not need to implement these traits since they are supported implicitly.
"Leaf" type refers to types which directly serialize like plain data types.
[`SerializeAs`] / [`DeserializeAs`] is very important for collection types, like `Vec` or `BTreeMap`, since they need special handling for they key/value de/serialization such that the conversions can be done on the key/values.
You also find them implemented on the conversion types, such as the [`DisplayFromStr`] type.
These make up the bulk of this crate and allow you to perform all the nice conversions to [hex strings], the [bytes to string converter], or [duration to UNIX epoch].

### Re-exporting `serde_as`

If `serde_as` is being used in a context where the `serde_with` crate is not available from the root
path, but is re-exported at some other path, the `crate = "..."` attribute argument should be used
to specify its path. This may be the case if `serde_as` is being used in a procedural macro -
otherwise, users of that macro would need to add `serde_with` to their own Cargo manifest.

The `crate` argument will generally be used in conjunction with [`serde`'s own `crate` argument].

For example, a type definition may be defined in a procedural macro:

```rust,ignore
// some_other_lib_derive/src/lib.rs

use proc_macro::TokenStream;
use quote::quote;

#[proc_macro]
pub fn define_some_type(_item: TokenStream) -> TokenStream {
    let def = quote! {
        #[serde(crate = "::some_other_lib::serde")]
        #[::some_other_lib::serde_with::serde_as(crate = "::some_other_lib::serde_with")]
        #[derive(::some_other_lib::serde::Deserialize)]
        struct Data {
            #[serde_as(as = "_")]
            a: u32,
        }
    };

    TokenStream::from(def)
}
```

This can be re-exported through a library which also re-exports `serde` and `serde_with`:

```rust,ignore
// some_other_lib/src/lib.rs

pub use serde;
pub use serde_with;
pub use some_other_lib_derive::define_some_type;
```

And the procedural macro can be used by other crates without any additional imports:

```rust,ignore
// consuming_crate/src/main.rs

some_other_lib::define_some_type!();
```

## De/Serialize Implementations Available

### Bytes / `Vec<u8>` to hex string

[`Hex`]

Requires the `hex` feature.

```ignore
// Rust
#[serde_as(as = "serde_with::hex::Hex")]
value: Vec<u8>,

// JSON
"value": "deadbeef",
```

### `Default` from `null`

[`DefaultOnNull`]

```ignore
// Rust
#[serde_as(as = "DefaultOnNull")]
value: u32,
#[serde_as(as = "DefaultOnNull<DisplayFromStr>")]
value2: u32,

// JSON
"value": 123,
"value2": "999",

// Deserializes null into the Default value, i.e.,
null => 0
```

### De/Serialize with `FromStr` and `Display`

Useful if a type implements `FromStr` / `Display` but not `Deserialize` / `Serialize`.

[`DisplayFromStr`]

```ignore
// Rust
#[serde_as(as = "serde_with::DisplayFromStr")]
value: u128,
#[serde_as(as = "serde_with::DisplayFromStr")]
mime: mime::Mime,

// JSON
"value": "340282366920938463463374607431768211455",
"mime": "text/*",
```

### `Duration` as seconds

[`DurationSeconds`]

```ignore
// Rust
#[serde_as(as = "serde_with::DurationSeconds<u64>")]
value: Duration,

// JSON
"value": 86400,
```

[`DurationSecondsWithFrac`] supports subsecond precision:

```ignore
// Rust
#[serde_as(as = "serde_with::DurationSecondsWithFrac<f64>")]
value: Duration,

// JSON
"value": 1.234,
```

Different serialization formats are possible:

```ignore
// Rust
#[serde_as(as = "serde_with::DurationSecondsWithFrac<String>")]
value: Duration,

// JSON
"value": "1.234",
```

The same conversions are also implemented for [`chrono::Duration`] with the `chrono` feature.

### Ignore deserialization errors

Check the documentation for [`DefaultOnError`].

### `Maps` to `Vec` of tuples

```ignore
// Rust
#[serde_as(as = "Vec<(_, _)>")]
value: HashMap<String, u32>, // also works with BTreeMap

// JSON
"value": [
    ["hello", 1],
    ["world", 2]
],
```

The [inverse operation](#vec-of-tuples-to-maps) is also available.

### `NaiveDateTime` like UTC timestamp

Requires the `chrono` feature.

```ignore
// Rust
#[serde_as(as = "chrono::DateTime<chrono::Utc>")]
value: chrono::NaiveDateTime,

// JSON
"value": "1994-11-05T08:15:30Z",
                             ^ Pretend DateTime is UTC
```

### `None` as empty `String`

[`NoneAsEmptyString`]

```ignore
// Rust
#[serde_as(as = "serde_with::NoneAsEmptyString")]
value: Option<String>,

// JSON
"value": "", // converts to None

"value": "Hello World!", // converts to Some
```

### Timestamps as seconds since UNIX epoch

[`TimestampSeconds`]

```ignore
// Rust
#[serde_as(as = "serde_with::TimestampSeconds<i64>")]
value: SystemTime,

// JSON
"value": 86400,
```

[`TimestampSecondsWithFrac`] supports subsecond precision:

```ignore
// Rust
#[serde_as(as = "serde_with::TimestampSecondsWithFrac<f64>")]
value: SystemTime,

// JSON
"value": 1.234,
```

Different serialization formats are possible:

```ignore
// Rust
#[serde_as(as = "serde_with::TimestampSecondsWithFrac<String>")]
value: SystemTime,

// JSON
"value": "1.234",
```

The same conversions are also implemented for [`chrono::DateTime<Utc>`] and [`chrono::DateTime<Local>`] with the `chrono` feature.

### Value into JSON String

Some JSON APIs are weird and return a JSON encoded string in a JSON response

[`JsonString`]

Requires the `json` feature.

```ignore
// Rust
#[derive(Deserialize, Serialize)]
struct OtherStruct {
    value: usize,
}

#[serde_as(as = "serde_with::json::JsonString")]
value: OtherStruct,

// JSON
"value": "{\"value\":5}",
```

### `Vec` of tuples to `Maps`

```ignore
// Rust
#[serde_as(as = "HashMap<_, _>")] // also works with BTreeMap
value: Vec<String, u32>,

// JSON
"value": {
    "hello": 1,
    "world": 2
},
```

The [inverse operation](#maps-to-vec-of-tuples) is also available.

[`chrono::DateTime<Local>`]: chrono_crate::DateTime
[`chrono::DateTime<Utc>`]: chrono_crate::DateTime
[`chrono::Duration`]: https://docs.rs/chrono/latest/chrono/struct.Duration.html
[`DefaultOnError`]: crate::DefaultOnError
[`DefaultOnNull`]: crate::DefaultOnNull
[`DeserializeAs`]: crate::DeserializeAs
[`DisplayFromStr`]: crate::DisplayFromStr
[`DurationSeconds`]: crate::DurationSeconds
[`DurationSecondsWithFrac`]: crate::DurationSecondsWithFrac
[`Hex`]: crate::hex::Hex
[`JsonString`]: crate::json::JsonString
[`NoneAsEmptyString`]: crate::NoneAsEmptyString
[`SerializeAs`]: crate::SerializeAs
[bytes to string converter]: crate::BytesOrString
[duration to UNIX epoch]: crate::DurationSeconds
[hex strings]: crate::hex::Hex
[serde_with#185]: https://github.com/jonasbb/serde_with/issues/185
[`serde`'s own `crate` argument]: https://serde.rs/container-attrs.html#crate