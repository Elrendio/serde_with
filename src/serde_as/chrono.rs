use crate::serde_as::*;
use chrono_crate::{DateTime, NaiveDateTime, Utc};
use serde::{Deserialize, Deserializer, Serialize, Serializer};

impl SerializeAs<NaiveDateTime> for DateTime<Utc> {
    fn serialize_as<S>(source: &NaiveDateTime, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let datetime = DateTime::<Utc>::from_utc(*source, Utc);
        datetime.serialize(serializer)
    }
}

impl<'de> DeserializeAs<'de, NaiveDateTime> for DateTime<Utc> {
    fn deserialize_as<D>(deserializer: D) -> Result<NaiveDateTime, D::Error>
    where
        D: Deserializer<'de>,
    {
        DateTime::<Utc>::deserialize(deserializer).map(|datetime| datetime.naive_utc())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::{collections::BTreeMap, str::FromStr};

    #[test]
    fn chrono_crate() {
        #[derive(Debug, Serialize, Deserialize, PartialEq)]
        pub struct SomeTime {
            #[serde(with = "As::<DateTime<Utc>>")]
            stamp: NaiveDateTime,
        }

        assert_eq!(
            serde_json::to_string(&SomeTime {
                stamp: NaiveDateTime::from_str("1994-11-05T08:15:30").unwrap()
            })
            .unwrap(),
            "{\"stamp\":\"1994-11-05T08:15:30Z\"}"
        );

        assert_eq!(
            SomeTime {
                stamp: NaiveDateTime::from_str("1994-11-05T08:15:30").unwrap()
            },
            serde_json::from_str("{\"stamp\":\"1994-11-05T08:15:30Z\"}").unwrap(),
        );
    }

    #[test]
    fn chrono_crate_opt() {
        #[derive(Debug, Serialize, Deserialize, PartialEq)]
        pub struct SomeTime {
            #[serde(with = "As::<Option<DateTime<Utc>>>")]
            stamp: Option<NaiveDateTime>,
        }

        assert_eq!(
            serde_json::to_string(&SomeTime {
                stamp: NaiveDateTime::from_str("1994-11-05T08:15:30").ok()
            })
            .unwrap(),
            "{\"stamp\":\"1994-11-05T08:15:30Z\"}"
        );

        assert_eq!(
            SomeTime {
                stamp: NaiveDateTime::from_str("1994-11-05T08:15:30").ok()
            },
            serde_json::from_str("{\"stamp\":\"1994-11-05T08:15:30Z\"}").unwrap(),
        );
    }

    #[test]
    fn chrono_crate_opt_vec() {
        #[derive(Debug, Serialize, Deserialize, PartialEq)]
        pub struct SomeTime {
            #[serde(with = "As::<Vec<Option<DateTime<Utc>>>>")]
            stamps: Vec<Option<NaiveDateTime>>,
        }

        assert_eq!(
            serde_json::to_string(&SomeTime {
                stamps: vec![
                    NaiveDateTime::from_str("1994-11-05T08:15:30").ok(),
                    NaiveDateTime::from_str("1994-11-05T08:15:31").ok()
                ],
            })
            .unwrap(),
            "{\"stamps\":[\"1994-11-05T08:15:30Z\",\"1994-11-05T08:15:31Z\"]}"
        );

        assert_eq!(
            SomeTime {
                stamps: vec![
                    NaiveDateTime::from_str("1994-11-05T08:15:30").ok(),
                    NaiveDateTime::from_str("1994-11-05T08:15:31").ok()
                ],
            },
            serde_json::from_str(
                "{\"stamps\":[\"1994-11-05T08:15:30Z\",\"1994-11-05T08:15:31Z\"]}"
            )
            .unwrap(),
        );
    }

    #[test]
    fn chrono_crate_hash_map() {
        #[derive(Debug, Serialize, Deserialize, PartialEq)]
        pub struct SomeTime {
            #[serde(with = "As::<BTreeMap<SameAs<i32>, DateTime<Utc>>>")]
            stamps: BTreeMap<i32, NaiveDateTime>,
        }

        assert_eq!(
            serde_json::to_string(&SomeTime {
                stamps: [
                    (1, NaiveDateTime::from_str("1994-11-05T08:15:30").unwrap()),
                    (2, NaiveDateTime::from_str("1994-11-05T08:15:31").unwrap()),
                ]
                .iter()
                .cloned()
                .collect(),
            })
            .unwrap(),
            "{\"stamps\":{\"1\":\"1994-11-05T08:15:30Z\",\"2\":\"1994-11-05T08:15:31Z\"}}"
        );

        assert_eq!(
            SomeTime {
                stamps: [
                    (1, NaiveDateTime::from_str("1994-11-05T08:15:30").unwrap()),
                    (2, NaiveDateTime::from_str("1994-11-05T08:15:31").unwrap()),
                ]
                .iter()
                .cloned()
                .collect(),
            },
            serde_json::from_str(
                "{\"stamps\":{\"1\":\"1994-11-05T08:15:30Z\",\"2\":\"1994-11-05T08:15:31Z\"}}"
            )
            .unwrap(),
        );
    }
}
