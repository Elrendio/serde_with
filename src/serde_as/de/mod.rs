use super::*;

pub(in crate::serde_as) mod impls;

pub trait DeserializeAs<'de, T>: Sized {
    fn deserialize_as<D>(deserializer: D) -> Result<T, D::Error>
    where
        D: Deserializer<'de>;

    // TODO: deserialize_as_into
}
