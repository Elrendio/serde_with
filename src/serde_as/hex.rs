use super::*;

#[derive(Copy, Clone, Debug, Default)]
pub struct Hex;

// TODO: AsRef
impl SerializeAs<Vec<u8>> for Hex {
    fn serialize_as<S>(source: &Vec<u8>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_str(&::hex::encode(source))
    }
}

impl<'de> DeserializeAs<'de, Vec<u8>> for Hex {
    fn deserialize_as<D>(deserializer: D) -> Result<Vec<u8>, D::Error>
    where
        D: Deserializer<'de>,
    {
        // FIXME: map decode errors
        <&'de str as Deserialize<'de>>::deserialize(deserializer).map(|s| ::hex::decode(s).unwrap())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hex_vec() {
        #[derive(Debug, Serialize, Deserialize, PartialEq)]
        pub struct SomeBytes {
            #[serde(with = "As::<Vec<Hex>>")]
            bytes: Vec<Vec<u8>>,
        }

        assert_eq!(
            serde_json::to_string(&SomeBytes {
                bytes: vec![vec![0, 1, 2, 13], vec![14, 5, 6, 7]]
            })
            .unwrap(),
            "{\"bytes\":[\"0001020d\",\"0e050607\"]}"
        );

        assert_eq!(
            SomeBytes {
                bytes: vec![vec![0, 1, 2, 13], vec![14, 5, 6, 7]]
            },
            serde_json::from_str("{\"bytes\":[\"0001020d\",\"0e050607\"]}").unwrap(),
        );
    }
}
