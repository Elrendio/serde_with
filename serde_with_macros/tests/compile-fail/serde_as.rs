use serde::Serialize;
use serde_with_macros::serde_as;

/// Test error message for conflicting *_as-annotations
#[serde_as]
#[derive(Serialize)]
struct ConflictingAsAnnotations {
    #[serde_as(as = "_", deserialize_as = "_")]
    a: u32,
    #[serde_as(as = "_", serialize_as = "_")]
    b: u32,
    #[serde_as(as = "_", deserialize_as = "_", serialize_as = "_")]
    c: u32,
}

/// Test error message for conflicts with serde's with-annotation
#[serde_as]
#[derive(Serialize)]
struct ConflictingWithAnnotations {
    #[serde_as(as = "_")]
    #[serde(with = "u32")]
    a: u32,
    #[serde_as(as = "_")]
    #[serde(deserialize_with = "u32")]
    b: u32,
    #[serde_as(as = "_")]
    #[serde(serialize_with = "u32")]
    c: u32,

    #[serde_as(deserialize_as = "_")]
    #[serde(with = "u32")]
    d: u32,
    #[serde_as(deserialize_as = "_")]
    #[serde(deserialize_with = "u32")]
    e: u32,
    #[serde_as(deserialize_as = "_")]
    #[serde(serialize_with = "u32")]
    f: u32,

    #[serde_as(serialize_as = "_")]
    #[serde(with = "u32")]
    g: u32,
    #[serde_as(serialize_as = "_")]
    #[serde(deserialize_with = "u32")]
    h: u32,
    #[serde_as(serialize_as = "_")]
    #[serde(serialize_with = "u32")]
    i: u32,
}

fn main() {}
