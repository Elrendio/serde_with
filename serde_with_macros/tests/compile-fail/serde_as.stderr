error: Cannot combine `as` with `deserialize_as`. Use `serialize_as` to specify different serialization code.
 --> $DIR/serde_as.rs:9:5
  |
9 |     a: u32,
  |     ^

error: Cannot combine `as` with `serialize_as`. Use `deserialize_as` to specify different deserialization code.
  --> $DIR/serde_as.rs:11:5
   |
11 |     b: u32,
   |     ^

error: Cannot combine `as` with `deserialize_as`. Use `serialize_as` to specify different serialization code.
  --> $DIR/serde_as.rs:13:5
   |
13 |     c: u32,
   |     ^

error: Cannot combine `serde_as` with serde's `with`, `deserialize_with`, or `serialize_with`.
  --> $DIR/serde_as.rs:21:5
   |
21 |     #[serde(with = "u32")]
   |     ^

error: Cannot combine `serde_as` with serde's `with`, `deserialize_with`, or `serialize_with`.
  --> $DIR/serde_as.rs:24:5
   |
24 |     #[serde(deserialize_with = "u32")]
   |     ^

error: Cannot combine `serde_as` with serde's `with`, `deserialize_with`, or `serialize_with`.
  --> $DIR/serde_as.rs:27:5
   |
27 |     #[serde(serialize_with = "u32")]
   |     ^

error: Cannot combine `serde_as` with serde's `with`, `deserialize_with`, or `serialize_with`.
  --> $DIR/serde_as.rs:31:5
   |
31 |     #[serde(with = "u32")]
   |     ^

error: Cannot combine `serde_as` with serde's `with`, `deserialize_with`, or `serialize_with`.
  --> $DIR/serde_as.rs:34:5
   |
34 |     #[serde(deserialize_with = "u32")]
   |     ^

error: Cannot combine `serde_as` with serde's `with`, `deserialize_with`, or `serialize_with`.
  --> $DIR/serde_as.rs:37:5
   |
37 |     #[serde(serialize_with = "u32")]
   |     ^

error: Cannot combine `serde_as` with serde's `with`, `deserialize_with`, or `serialize_with`.
  --> $DIR/serde_as.rs:41:5
   |
41 |     #[serde(with = "u32")]
   |     ^

error: Cannot combine `serde_as` with serde's `with`, `deserialize_with`, or `serialize_with`.
  --> $DIR/serde_as.rs:44:5
   |
44 |     #[serde(deserialize_with = "u32")]
   |     ^

error: Cannot combine `serde_as` with serde's `with`, `deserialize_with`, or `serialize_with`.
  --> $DIR/serde_as.rs:47:5
   |
47 |     #[serde(serialize_with = "u32")]
   |     ^
